### PR TITLE
[2.5] fix gouroutine leak when sharedcontroller has been started

### PR DESCRIPTION
**Pull request by JacieChao**
1. Rancher will start a usermanager controller for each downstream cluster, and will start management controller factory [here](https://github.com/rancher/rancher/blob/08ad9f793a2345478a26a514fca021c8e648f2f3/pkg/types/config/context.go#L476). The management controller factory will start only once cause it checked `started` state [here](https://github.com/rancher/lasso/blob/release/v2.5.8-patch1/pkg/controller/controller.go#L143). If there's something wrong with downstream cluster and usermanager controller has to stop, it canceled the context and skip goroutine [here](https://github.com/rancher/lasso/blob/release/v2.5.8-patch1/pkg/controller/sharedcontroller.go#L101). The management controller factory is never canceled, so that each time we start a usermanager controller, the goroutine leaks [here](https://github.com/rancher/lasso/blob/release/v2.5.8-patch1/pkg/controller/sharedcontroller.go#L97). So we also need to check `started` state to avoid create more goroutine leaks if controller has started.

2. It set `started` state in Start() function but never set `unstarted` state for shutting down works when channel done. The controller may fail to start when we restart it.

### CODE REVIEW

**Comment by deniseschannon** [comment]
@JacieChao can you make a PR to master? 
cc: @niusmallnan

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -133,6 +133,7 @@ func (c *controller) run(workers int, stopCh <-chan struct{}) {
 	}
 
 	<-stopCh
+	c.started = false
```
Comment (ibuildthecloud) [inline pkg/controller/controller.go:136]: You need to acuire the startLock before writing this.

Comment (niusmallnan) [reply pkg/controller/controller.go:136]: @JacieChao It seems like it should be:

```
c.startLock.Lock()
defer c.startLock.Unlock()
c.started = false
```

Comment (JacieChao) [reply pkg/controller/controller.go:138]: fixed

**Comment by JacieChao** [comment]
@deniseschannon PR to master submitted https://github.com/rancher/lasso/pull/12

**Pull request merged**